operator/v1alpha1/types_etcdbackup.go

```diff
@@ -0,0 +1,102 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:resource:scope=Cluster
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+//
+// # EtcdBackup provides configuration options and status for a one-time backup attempt of the etcd cluster
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type EtcdBackup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec EtcdBackupSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status EtcdBackupStatus `json:"status"`
+}
+
+type EtcdBackupSpec struct {
+	// PVCName specifies the name of the PersistentVolumeClaim which binds a PersistentVolume where the
+	// etcd backup file would be saved
+	// +kubebuilder:validation:Required
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="pvcName is immutable once set"
+	PVCName string `json:"pvcName"`
```

Comment:
Is this actually configurable in some sense? A cluster-admin shouldn't be allowed to create PVCs in your namespace, right?

Reply:
Forgot to mention that this PVC should be in the "cluster-etcd-operator" namespace.
And yes I'm assuming that the cluster-admin does have the permissions to create PVCs in our namespace (or any namespace I think).

But please clarify if that's not the case, or if it's a bad idea to allow PVCs in the etcd-operator namespace.

Or did you mean that we should allow the admin to choose their own namespace for the PVC so they don't have to use the "cluster-etcd-operator" namespace?

Reply:
the backup job must run in the `openshift-etcd` namespace: https://github.com/openshift/cluster-etcd-operator/blob/master/bindata/etcd/cluster-backup-pod.yaml#L5

so the PVC would also need to go there. Otherwise I assume we need to create it?

Reply:
Sharing our operator or operand namespace with a user strikes me as unusual and rules like immutability seem even stranger.  What does letting a customer provide their PVC allow us to do?

Could they provide a PV (cluster scoped) and we provide the PVC?

Reply:
So my thought process on why I have the admin providing their PVC here:
- The admin is expected to provide their own PV (or storage class for dynamic provisioning) so they have control over how to provision the storage
- The admin is also expected to create their own PVC so they can choose the pvc spec: storage class, access modes, resource requests, etc
- The operator can’t own the PVC lifecycle, since deleting the PVC also triggers the deletion or reclamation of the PV.
  - Plus if the PV is the same for every subsequent `EtcdBackup` request, the operator would have to reuse the same PVC, since you can’t bind different PVCs to a single PV. But if it’s a different PV name, or a dynamic provisioning storage class then the operator has to use a new PVC.
  - So for multiple reasons the PVC lifecycle seems best left to the admin.
- Since the backup execution job/pod has to reference the PVC in its own namespace that means the PVC has to be in the namespace of our choosing e.g `openshift-etcd`. So that’s not configurable.
- So the expected workflow then is to allow the admin to create their PVC and then point us to that in the `EtcdBackup` spec 
- The reason I have `spec.pvcName` as immutable is because the `EtcdBackup` CR represents a one-time backup execution, and changing the `pvcName` after the CR creation from `pvc1` to `pvc2` means we would have to redo an in progress or completed backup and save it again to wherever `pvc2` is bound.  And that goes against the “1 `EtcdBackup` == 1 backup attempt” design. 

So that’s why I have the `pvcName` as configurable on creation but immutable after.
But happy to hear different ideas on if there's something simpler or I’m missing something.

If sharing our namespace with the admin is an absolute no-go then I need to rethink how we could get the pvc and our backup pod in the same namespace without us owning the PVC lifecycle.

Reply:
The backup pod running the backup script [needs to run as privileged](https://github.com/openshift/cluster-etcd-operator/blob/c7cc2709d2d5b81f55a72a8ddd839ef0115eac96/bindata/etcd/cluster-backup-pod.yaml#L44C5-L45C23) so it can grab the static pod manifests and certs from the host filesystem.

We could run it outside `openshift-etcd`, in a namespace of the admin's choice, provided it's setup to allow privileged pods, i.e the ns has the following labels like `openshift-etcd` does:
```yaml
pod-security.kubernetes.io/audit: privileged
pod-security.kubernetes.io/enforce: privileged
pod-security.kubernetes.io/warn: privileged
```
So that way the admin provides us with the name and namespace of their PVC in the `EtcdBackup` spec and we can use that to execute and save the backup.

@deads2k WDYT? Would that address the namespace sharing concern?

Reply:
I had to go back and read the docs. Because of the double-opt-in for PV and PVC (justified decision), I'm convinced on the acceptability of requiring a PVC, **but** we need a zero-config option that will take backups and store them locally on the masters. I've suggested a mechanism in a comment above to handle this.